ci/tests/puppeteer/scenarios/oidc-authzcode-pac4j-delegated-login/script.js

```diff
@@ -0,0 +1,37 @@
+const puppeteer = require('puppeteer');
+const assert = require('assert');
+const cas = require('../../cas.js');
+
+(async () => {
+    const browser = await puppeteer.launch(cas.browserOptions());
+    const page = await browser.newPage();
+    const url = "https://localhost:8443/cas/oidc/authorize?response_type=code&client_id=client&scope=openid%20MyCustomScope&redirect_uri=http://localhost:8080/oidclogin&nonce=3d3a7457f9ad3&state=1735fd6c43c14&claims=%7B%22userinfo%22%3A%20%7B%20%22name%22%3A%20%7B%22essential%22%3A%20true%7D%2C%22phone_number%22%3A%20%7B%22essential%22%3A%20true%7D%7D%7D";
+    await page.goto(url);
+
+    await cas.assertVisibility(page, '#loginProviders')
+    await cas.assertVisibility(page, 'li #DelegateLogin')
+    await cas.click(page, "li #DelegateLogin")
+    await page.waitForNavigation();
+
+    await cas.loginWith(page, "casuser", "Mellon");
+
+    await page.waitForTimeout(2000)
+
+    await cas.assertVisibility(page, '#userInfoClaims')
+
+    await cas.assertVisibility(page, '#scopes')
+
+    await cas.assertVisibility(page, '#MyCustomScope')
+
+    await cas.assertVisibility(page, '#openid')
+
+    await cas.assertVisibility(page, '#informationUrl')
+
+    await cas.assertVisibility(page, '#privacyUrl')
+
+    await cas.assertVisibility(page, '#name')
+
+    await cas.assertVisibility(page, '#phone_number')
```

Comment:
The test has no assertion to verify the final sequence. You need to assert the fact you either get or do not get a service ticket.

Reply:
> Actual result: http://localhost:8080/oidclogin receives the service ticket that was supposed for CAS's callback endpoint.

This needs to be asserted.

Reply:
It is asserted implicitly by validating the attributes on the target page. The target page will be unable to validate the token, because it is a service ticket.

Reply:
Sorry I don't follow. What attributes? which target page?

Reply:
To my understanding, after a successful authentication, CAS redirects the request back to redirect_uri=http://localhost:8080/oidclogin where the authorization code is validated against CAS and the released user profile is shown. This is what is validated by

`await cas.assertVisibility(page, '#userInfoClaims')`

etc., right?

Reply:
Correct, but no :) `userInfoClaims` is an element that is shown on CAS OIDC consent page where the browser is asking the user to authorize the release of claims. In your test, you are still on the CAS consent page, and have not moved onto the target application.

What you need to do is:

1. Click the allow button the page and allow CAS to make that redirect
2. Examine the target application and make sure you are in the right place, depending on the output you get.
3. Check to see if you have a ticket in the URL when you land on the target application.


Did you ever have a look at the existing tests? You seem to be duplicating this, no?
https://github.com/apereo/cas/blob/master/ci/tests/puppeteer/scenarios/pac4j-delegated-login-oidc-rp-cas-idp/script.js

Reply:
PS Also, speaking in broad terms, tests should never be "implicit". If you look at something 6 months from now, it should be obvious what the test is doing and/or testing. If you find this to be false in other areas, don't let those serve as an example. Let's fix those (later) to be explicit too.

And of course, thank you for looking into this with an automated test. Much appreciated.

Reply:
I see, sorry for the confusion. "Be right back"... ;)

Reply:
You are right, I did not take that test into consideration because it was using delegation via cas-protocol and it was not failing so far. The new test is now "successfully failing" with a Service Ticket in the OIDC response.

Reply:
Hmm wait, something is still weird with the test...

Is there any chance to get SAML2 or OIDC delegation to work with the demo CAS at cas.herokuapp.com? Would make these tests a lot easier...

Reply:
After merging the lastest changes from the master into my test branch, I can no longer reproduce the behaviour described in PR 5166. I assume your latest changes have fixed the problem in another way. I will come back to you if I encounter the problem again!